File: Problem1609/solution.cpp
/* 
Let n be the number of nodes in the tree and h be the height of the tree

Time Complexity: O(n)
Space Complexity: O(n) -> values vector use O(h) but the worst case in call stack is O(n)

A recursive solution using dfs.
*/

#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void solver(TreeNode* node, int level = 0){
        int sign = -2*(level%2)+1;
        if(node == nullptr || solution == false){return;}
        if(node->val%2 == level%2){
            solution = false;
            return;
        }
        if(values.size() <= level){values.push_back(node->val);}
        else if(values[level]*sign < node->val*sign){values[level] = node->val;}
        else{solution = false;}
        solver(node->left, level+1);
        solver(node->right, level+1);
    }
    bool isEvenOddTree(TreeNode* root) {
        solver(root);
        return solution;
    }
private:
    bool solution = true;
    std::vector<int> values = {};
};

int main(){

    Solution sol = Solution();
    TreeNode* root = new TreeNode(5, new TreeNode(4, new TreeNode(3), new TreeNode(5)), new TreeNode(2, new TreeNode(7), nullptr));

    cout << sol.isEvenOddTree(root) << endl;

    delete root->right;
    delete root->left;
    delete root;

}

File: Problem1609/solution.py
""" 
Let n be the number of nodes in the tree and h be the height of the tree

Time Complexity: O(n)
Space Complexity: O(n) -> values vector use O(h) but the worst case in call stack is O(n)

A recursive solution using dfs.

"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    solution:bool = True

    def isEvenOddTree(self, root: TreeNode) -> bool:
        values:list = []

        def solver(node: TreeNode, level:int = 0) -> None:
            sign:int = -2*(level%2)+1
            if node is None or self.solution == False:
                return
            if node.val%2 == level%2:
                self.solution = False
                return
            if len(values) <= level:
                values.append(node.val)
            elif values[level]*sign < node.val*sign:
                values[level] = node.val
            else:
                self.solution = False
            solver(node.left, level+1)
            solver(node.right, level+1)

        solver(root)
        return self.solution
    
sol:Solution = Solution()
root:TreeNode = TreeNode(5, TreeNode(4, TreeNode(3), TreeNode(5)), TreeNode(2, TreeNode(7), None))
print(sol.isEvenOddTree(root))